### PR TITLE
Makes library installable.

### PR DESCRIPTION
**Pull request by jessezach**
- Added parallesisation.
- Moved classes to respective modules
- Using opts parse to parse options

How to install and use:
cd to root and run `python setup.py install`

To generate report: 
`robotmetrics [:options]`

Rest of the usage commands are same

How to test and contribute:
cd to root and run `python setup.py develop`

Note
All options now begin with -- (double hyphens)

### CODE REVIEW

**Comment by jz-jess** [comment]
@adiralashiva8 I have changed the directory structure and renamed robotframework-metrics to robotframework_metrics to avoid installing errors. Added parallelisation as well.
I have tested at my end. Seems to be working fine. However it might still require testing with large files and negative testing with/without options

**Comment by adiralashiva8** [comment]
@jz-jess Changes looks awesome but I have few minor concerns which need you help

Robotframework-metrics have some customization stuff like

1. Adding own image
2. Adding own links
3. Ignoring few library keywords
4. Specifying email recipients

I feel recent changes (Makes library installable) wont support this features. Is there any possibilities to support this features?

**Comment by jz-jess** [comment]
@adiralashiva8 I missed it. Supporting this shouldn't be a problem. However opening the file and editing it is not the way to go.
We already have some default values in the code. User should be able to override it by either passing it as a command line param or store in a config (yaml, etc) file and pass the path as a param. Does this make sense to you?

**Comment by adiralashiva8** [comment]
> @adiralashiva8 I missed it. Supporting this shouldn't be a problem. However opening the file and editing it is not the way to go.
> We already have some default values in the code. User should be able to override it by either passing it as a command line param or store in a config (yaml, etc) file and pass the path as a param. Does this make sense to you?

I like suggested approach (command line).

**Comment by jz-jess** [comment]
I'll make the changes, meanwhile do let me know if you're able to find any bugs

**Comment by adiralashiva8** [comment]
> I'll make the changes, meanwhile do let me know if you're able to find any bugs

@jz-jess I was not able to install library because Installing gevent is failing in my machine. May be users also going to face this issue (gevent need to be specific version based on python version )

What I thinking is lets drop from performance improvement and as part of this pull request focus on making library installable if you agree?

**Comment by jz-jess** [comment]
@adiralashiva8 Will try with the standard multiprocess library then. Will remove if it doesn't work as expected

**Comment by jz-jess** [comment]
@adiralashiva8 Can you tell me the error you get when you install gevent in your system? Also your OS details. Gevent supports python 2.7 and above.

**Comment by adiralashiva8** [comment]
> @adiralashiva8 Will try with the standard multiprocess library then. Will remove if it doesn't work as expected. Let me remove it for now to enable testing

@jz-jess After removing gevent I was able to install and generate report. I found two issue (1 can be ignored)

**Issues:**

1. Email Metrics: When user generate email (within report) .eml file showing html content (see here for working sample - https://robotframework-metrics.netlify.com/#)
2. Generate Report: By default email will be sent whenever robotmetrics command is executed (default should be true) - I'm okay with this

**Run commands I used**
`robotmetrics --output ouput1.xml --email true`

**Comment by adiralashiva8** [comment]
> @adiralashiva8 Can you tell me the error you get when you install gevent in your system? Also your OS details. Gevent supports python 2.7 and above.

Command: `pip install gevent`
Error: `Command "c:\users\shiva.adirala\appdata\local\programs\python\python37-32\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\SHIVA~1.ADI\\AppData\\Local\\Temp\\pip-install-gxqqogaq\\gevent\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\SHIVA~1.ADI\AppData\Local\Temp\pip-record-2e0c8llb\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\SHIVA~1.ADI\AppData\Local\Temp\pip-install-gxqqogaq\gevent\`

OS: Windows 10
Python: 3.7.0

**Comment by jz-jess** [comment]
@adiralashiva8 Can you try installing gevent from this link?
https://github.com/gevent/gevent/downloads

This has binary for windows. Let me know if this works. If it does this link can be added in the ReadMe.

**Comment by jz-jess** [comment]
https://stackoverflow.com/questions/40018405/cython-cannot-open-include-file-io-h-no-such-file-or-directory

**Comment by adiralashiva8** [comment]
> https://stackoverflow.com/questions/40018405/cython-cannot-open-include-file-io-h-no-such-file-or-directory

May be **Download visual studio build tools and install** solution works but I cant ask users to download all this stuff for generating simple metrics report :). What you feel

**Comment by jz-jess** [comment]
@adiralashiva8 I have multiple approaches as of now:

1. Add this as optional in readme. Will import gevent within try except block. If it breaks we execute sequentially. Otherwise we pass it to a gevent group.

2. Do this using multiprocessing library. If there are no hiccups. We should be good to go.

3. Remove concurrency all together.

**Comment by adiralashiva8** [comment]
> @adiralashiva8 I have multiple approaches as of now:
> 
> 1. Add this as optional in readme. Will import gevent within try except block. If it breaks we execute sequentially. Otherwise we pass it to a gevent group.
> 2. Do this using multiprocessing library. If there are no hiccups. We should be good to go.
> 3. Remove concurrency all together.

I'm good with 1st approach 👍  (if users are concern with performance then they will use gevent else go with sequence)

**Comment by jz-jess** [comment]
Alright, will fix everything at night and comment.
Report looks very cool btw. We should think of ways to generate this on the fly than user having to generate it manually

**Comment by adiralashiva8** [comment]
> Alright, will fix everything at night and comment.
> Report looks very cool btw. We should think of ways to generate this on the fly than user having to generate it manually

Robotframework-metircs uses output.xml which will be generated after test/suite execution we don't have option to have ouput.xml on fly. There is other approach (using listeners) but i'm not aware on it.

Really excited for final changes and Eagerly waiting to download **robotframework-metrics** using pip command

**Comment by jz-jess** [comment]
> There is other approach (using listeners) but i'm not aware on it.

Once the library becomes installable we can use the `xunit_file` listener method to get the output xml and then create the html. In future this library can be an external listener and just run from command line with arguments like `robot --listener MyListener:arg1:arg2 tests.robot

**Comment by jz-jess** [comment]
@adiralashiva8 

> 1. Email Metrics: When user generate email (within report) .eml file showing html content (see here for working sample - https://robotframework-metrics.netlify.com/#)

What is the issue here? The report I generated seems to have email-metrics. It works same as the example you provided

**Comment by adiralashiva8** [comment]
When I opened downloaded email - HTML content is show in email body - let me recheck this

**Comment by jesse019** [comment]
> Need Discussion: In robotmetrics.py we have hardcoded ignore_library and ignore_type value, but In future I want to include hardcoded library keywords in report. How can we achieve this?

You can iterate through the variables and add them to the html text like you have for others.

File: robotframework_metrics/robotmetrics.py [hunk 1]
```diff
@@ -0,0 +1,1196 @@
...
+        'SeleniumLibrary',
+        'String',
+        'Collections',
+        'DateTime',
+        ]
+
+    if opts.ignore is None:
+        pass
+    else:
+        ignore_list = opts.ignore.split(',') if opts.ignore else ''
```
[the remark this answers is not in the document]
Comment (jesse019) [reply robotframework_metrics/robotmetrics.py]: I had done it last night. I guess it worked for me. Will test once more late evening fix and update code accordingly.

**Comment by jesse019** [comment]
@adiralashiva8 To publish to pip follow instructions here:
https://packaging.python.org/tutorials/packaging-projects/

Once everything is good to go. Give it a try. Let me know if you need help.

**Review by adiralashiva8 — APPROVED** [review]

**Pull request merged**